### ISSUE TITLE
kops-controller enters crashloop on update to k8s 1.19.7

### ISSUE THREAD

**Issue by MichaelJDrK** [kubernetes/kops#10704]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

Version 1.19.0 (git-04d36d7d92c72601efd918877fc180c846129ffb)

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-17T02:13:01Z", GoVersion:"go1.15.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.15", GitCommit:"73dd5c840662bb066a146d0871216333181f4b64", GitTreeState:"clean", BuildDate:"2021-01-13T13:14:05Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}

**3. What cloud provider are you using?**

AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

1. Create a new cluster with kops 1.19.0 and Kubernetes 1.18.15. (1 master, 2 nodes)
2. Set "kubernetesVersion: 1.19.7" in the cluster.yaml
3. Update cluster.yaml in S3: "kops replace -f cluster.yaml"
4. Update cloud resources: "kops update cluster $NAME --yes"

**5. What happened after the commands executed?**

This will trigger an update to the kops-controller deployment. The kops-controller will now be in a crashloop because of the new config.yaml. 

E0201 09:34:10.278898       1 main.go:107] setup "msg"="unable to start server" "error"="reading \"ca\" certificate: open /etc/kubernetes/kops-controller/pki/ca.pem: no such file or directory"                                   
 stream closed  

kops-controller expects certificates on the node, which are not yet available, because there wasn't a rolling update yet.
You cannot do a rolling-update, because the cluster validation will not pass (because of the crashloop)

**6. What did you expect to happen?**

The cluster should remain stable. kops-controller should not enter a crashloop because of the missing certificates. There should be some smooth transition.

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```yaml
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: null
  name: FOOBAR
spec:
  additionalPolicies:
    master: |
      [
        {
          "Effect": "Allow",
          "Action": ["sts:AssumeRole"],
          "Resource": "*"
        }        
      ]
  api:
    dns: {}
  authorization:
    rbac: {}
  channel: stable
  cloudProvider: aws
  configBase: s3://FOOBAR
  containerRuntime: docker
  dnsZone: FOOBAR
  etcdClusters:
  - cpuRequest: 200m
    etcdMembers:
    - instanceGroup: master-eu-central-1a
      name: a
      volumeSize: 3
    memoryRequest: 100Mi
    name: main
  - cpuRequest: 100m
    etcdMembers:
    - instanceGroup: master-eu-central-1a
      name: a
      volumeSize: 3
    memoryRequest: 100Mi
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubelet:
    anonymousAuth: false
    authenticationTokenWebhook: true
    authorizationMode: Webhook    
  kubeAPIServer:
    disableBasicAuth: true
    enableProfiling: false
  kubeControllerManager:
    terminatedPodGCThreshold: 10
    enableProfiling: false
  kubeDNS:
    provider: CoreDNS
  kubeScheduler:
    enableProfiling: false    
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.19.7
  masterPublicName: api.FOOBAR
  networkCIDR: 172.20.0.0/16
  networkID: FOOBAR
  networking:
    weave:
      mtu: 8912
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 0.0.0.0/0
  subnets:
  - cidr: 172.20.32.0/19
    id: FOOBAR
    name: eu-central-1a
    type: Public
    zone: eu-central-1a
  - cidr: 172.20.64.0/19
    id: FOOBAR
    name: eu-central-1b
    type: Public
    zone: eu-central-1b
  topology:
    dns:
      type: Public
    masters: public
    nodes: public

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: null
  labels:
    kops.k8s.io/cluster: FOOBAR
  name: master-eu-central-1a
spec:
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210119.1
  machineType: t3.medium
  maxPrice: "0.020"
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-eu-central-1a
  role: Master
  rootVolumeOptimization: true  
  rootVolumeSize: 8
  subnets:
  - eu-central-1a

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: null
  labels:
    kops.k8s.io/cluster: FOOBAR
  name: nodes-eu-central-1a
spec:
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210119.1
  machineType: t3.medium
  maxPrice: "0.020"
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: nodes-eu-central-1a
  role: Node
  rootVolumeOptimization: true
  rootVolumeSize: 10
  subnets:
  - eu-central-1a
  
---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: null
  labels:
    kops.k8s.io/cluster:FOOBAR
  name: nodes-eu-central-1b
spec:
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210119.1
  machineType: t3.medium
  maxPrice: "0.020"
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: nodes-eu-central-1b
  role: Node
  rootVolumeOptimization: true
  rootVolumeSize: 10
  subnets:
  - eu-central-1b  

```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**
-/-

**9. Anything else do we need to know?**
-/-

**Comment by MichaelJDrK** [comment]
Multi-master clusters are affected as well, same as private topology. Easily reproducable.

It seems, that it's NOT possible to update ANY cluster to 1.19.7 (at least not on AWS).

**Comment by rifelpet** [comment]
Hi @MichaelJDrK thanks for the report. Can you confirm if the kops-controller deployment has the `kops.k8s.io/kops-controller-pki: ""` nodeSelector? and does the master node have that label?

https://github.com/kubernetes/kops/blob/bdeec95a04af00faf6e2052f7d86f97cad48d83a/upup/models/cloudup/resources/addons/kops-controller.addons.k8s.io/k8s-1.16.yaml.template#L48

That nodeSelector is meant to prevent the new kops-controller pods from being scheduled onto old nodes that dont have the certificates in place.

**Comment by MichaelJDrK** [comment]
Yes, the kops-controller deployment and the master node both have the `kops.k8s.io/kops-controller-pki: ""` label. 

I did some more tests. What happens is:
Case a) kops 1.18.3, Cluster 1.18.15 
Update kops to 1.19.0, update cluster first (without updating to Kubernetes 1.19.7 in the same step).
-> new masters will get the `kops.k8s.io/kops-controller-pki: ""` label, even the certificates are not there. Not an issue yet
Update kubernetes to 1.19.7
-> kops-controller will update, run on the mastes and don't find certificates

This affects a two step upgrade of an exististing cluster (kops first, then kubernetes) as well as a fresh k8s 1.18.15 (installed with kops 1.19.0) update to k8s 1.19.7

Case b) kops 1.18.3, Cluster 1.18.15 
Update kops to 1.19.0 and Update Kubernetes to 1.19.7 in one step
-> kops-controller will update. The old pods will get terminated, but not scheduled on the master, because the label is missing. 
-> rolling-update works as expected

**Comment by rifelpet** [comment]
I think the problem is that Kops 1.19's protokube component will always add the `kops-controller-pki` label to control plane nodes, regardless of kubernetes version:

https://github.com/kubernetes/kops/blob/1885ccee1730d6b6b6326f5bdfaaf02cac19e74a/protokube/pkg/protokube/labeler.go#L49-L52

but the PKI files are only populated by nodeup for k8s >= 1.19:

https://github.com/kubernetes/kops/blob/1885ccee1730d6b6b6326f5bdfaaf02cac19e74a/nodeup/pkg/model/kops_controller.go#L41-L51

https://github.com/kubernetes/kops/blob/1885ccee1730d6b6b6326f5bdfaaf02cac19e74a/pkg/apis/kops/model/features.go#L24-L26

so the kops 1.19 version of the kops-controller pod gets incorrectly scheduled onto k8s < 1.19 control plane nodes.

It seems like we should not add  the `kops-controller-pki` label to nodes that are k8s < 1.19.

**Comment by roeera** [comment]
@rifelpet @hakman I am currently facing the exact problem.
I updated my cluster to 1.20.13.
1 master, 1 node.
Node can't connect to the master. From journalctl I see the following error message :
 `error running task "BootstrapClientTask/BootstrapClient" (9m51s remaining to succeed): Post "https://kops-controller.internal.sandbox-data.cluster.xx:3988/bootstrap": dial tcp 10.0.0.29:3988: connect: connection refused`
when listing the current pods in kube-system:
 kops-controller-6mx6n                               0/1     CrashLoopBackOff
On its log there is an error :
`etup "msg"="unable to start server" "error"="reading \"kubernetes-ca\" certificate: open /etc/kubernetes/kops-controller/pki/kubernetes-ca.crt: no such file or directory"`

Please help with this! How can I end this loop ?!

**Comment by cam3ron2** [comment]
@roeera I experienced the same issue when updating Kubernetes from 1.21.4 -> 1.21.8 and Kops from 1.21.1 -> 1.22.3

One master node upgraded successfully, but wound up still running kops-controller 1.21.1 instead of 1.22.3

The update hung itself on the second master node because kops-controller was crashlooped. 
Logs from the kops-controller on the affected node: 
`E0125 22:39:57.181115       1 deleg.go:144] setup "msg"="unable to start server" "error"="reading \"kubernetes-ca\" certificate: open /etc/kubernetes/kops-controller/pki/kubernetes-ca.pem: no such file or directory"`
Node version was still 1.21.4. The kops-controller pod on that node was somehow 1.22.3, others were still 1.21.1.

Rolling the pod on the affected node back to 1.21.1 did not fix the issue, but rolling back to 1.18.3 did. I picked 1.18.3 since it did not have the host volume `/etc/kubernetes/kops-controller/pki/` in the daemonset manifest.

From there I did another `kops update` and `kops rolling-update` which allowed me to update the second master node, but I then encountered the same problem on the 3rd master node. I once again did a `kubectl rollout undo` on the kops-controller daemonset to roll it back to 1.18.3 and deleted the pod, which allowed the upgrade to proceed again. 

Once all 3 masters were upgraded, node upgrades began failing because nodes would not join the cluster. I was able to then rollback the kops-controller daemonset to 1.22.3 and delete the pods to roll them out. Once that was done, the upgrade proceeded as normal.

**Comment by olemarkus** [comment]
We don't really monitor closed issues. Please open new issues. If things are actively broken, I recommend reaching out on slack (#kops-users on the kubernetes slack)

**Comment by dmytropetryk** [comment]
I had the same case when I upgraded kops from 1.20.3 to 1.21.5. K8s version was 1.20.10.  Updated only kops. After the update, kops-controller pods were in CrashLoopBackOff...
On its log there is an error :
etup "msg"="unable to start server" "error"="reading \"kubernetes-ca\" certificate: open /etc/kubernetes/kops-controller/pki/kubernetes-ca.crt: no such file or directory"
Fixed by applying manifest k8s-1.16.yaml from my S3 bucket kops state "--state=s3://" location path /addons/kops-controller.addons.k8s.io/k8s-1.16.yaml. Looks like not all resources weren't updated correctly after the normal Kops upgrade...
example: kubectl apply -f addons/kops-controller.addons.k8s.io/k8s-1.16.yaml
